From 8d9ac20c724ed6a35eb847c35c32f290d6100af5 Mon Sep 17 00:00:00 2001
From: Volodymyr Babchuk <volodymyr.babchuk@globallogic.com>
Date: Thu, 7 Jul 2016 20:52:43 +0300
Subject: [PATCH 3/4] arm/optee: disable SHM cache

OPTEE can cache SHM buffer on it's side. It is normal if there are only
one guest. But if there are number of guests, OPTEE will try to
use cached value when working with another guest. This will lead to
crash.

Signed-off-by: Volodymyr Babchuk <volodymyr.babchuk@globallogic.com>
---
 xen/arch/arm/optee/smc.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/xen/arch/arm/optee/smc.c b/xen/arch/arm/optee/smc.c
index d59ab91..e280595 100644
--- a/xen/arch/arm/optee/smc.c
+++ b/xen/arch/arm/optee/smc.c
@@ -141,6 +141,12 @@ int optee_handle_smc(struct cpu_user_regs *regs)
 	case OPTEE_SMC_GET_SHM_CONFIG:
 		do_process_get_shm_config(regs);
 		break;
+	case OPTEE_SMC_ENABLE_SHM_CACHE:
+		/* We can't allow guests to enable SHM cache */
+		/* as OPTEE can cache other guest's SHM      */
+		/* So, we will pretend that cache is enabled */
+		regs->x0 = OPTEE_SMC_RETURN_OK;
+		break;
 	default:
 		/* Just forward request to OPTEE */
 		execute_smc(regs);
-- 
2.8.3

